atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -0,0 +1,428 @@
+                if (startToken.compareTo(endToken) <= 0) {
+                    newTokenRing.put(Range.openClosed(startToken, endToken), hosts);
+                } else {
+                    // Handle wrap-around
+                    newTokenRing.put(Range.greaterThan(startToken), hosts);
+                    newTokenRing.put(Range.atMost(endToken), hosts);
+                }
+            }
+            tokenMap = newTokenRing.build();
+
+        } catch (Exception e) {
+            log.error("Couldn't grab new token ranges for token aware cassandra mapping!", e);
+        }
+    }
+
+    private FunctionCheckedException<Cassandra.Client, List<TokenRange>, Exception> describeRing = new FunctionCheckedException<Cassandra.Client, List<TokenRange>, Exception>() {
+        @Override
+        public List<TokenRange> apply (Cassandra.Client client) throws Exception {
+            return client.describe_ring(config.keyspace());
+        }};
+
+    public <V, K extends Exception> V runWithRetry(FunctionCheckedException<Cassandra.Client, V, K> f) throws K {
+       return runWithRetryOnHost(getRandomGoodHost().getHost(), f);
+    }
+
+    public <V, K extends Exception> V runWithRetryOnHost(InetSocketAddress specifiedHost, FunctionCheckedException<Cassandra.Client, V, K> f) throws K {
+        int numTries = 0;
+        while (true) {
+            CassandraClientPoolingContainer hostPool = currentPools.get(specifiedHost);
+
+            if (blacklistedHosts.containsKey(specifiedHost) || hostPool == null) {
+                log.warn("Randomly redirected a query intended for host %s because it was not currently a live member of the pool.", specifiedHost);
+                hostPool = getRandomGoodHost();
+            }
+
+            try {
+                return hostPool.runWithPooledResource(f);
+            } catch (Exception e) {
+                numTries++;
+                this.<K>handleException(numTries, hostPool.getHost(), e);
+            }
+        }
+    }
+
+    public <V, K extends Exception> V run(FunctionCheckedException<Cassandra.Client, V, K> f) throws K {
+        return runOnHost(getRandomGoodHost().getHost(), f);
+    }
+
+    public <V, K extends Exception> V runOnHost(InetSocketAddress specifiedHost, FunctionCheckedException<Cassandra.Client, V, K> f) throws K {
+        CassandraClientPoolingContainer hostPool = currentPools.get(specifiedHost);
+        return hostPool.runWithPooledResource(f);
+    }
+
+        @SuppressWarnings("unchecked")
+    private <K extends Exception> void handleException(int numTries, InetSocketAddress host, Exception e) throws K {
+        if (e instanceof CassandraClientFactory.ClientCreationFailedException
+                || e instanceof TTransportException
+                || e instanceof TimedOutException
+                || e instanceof SocketTimeoutException
+                || e instanceof UnavailableException
+                || e instanceof InsufficientConsistencyException) {
```

Comment:
Need to handle `NoSuchElementException`

Should probably extract this condition out to a separate method